### DISCUSSION TITLE
Request for More Detailed Documentation on MuSGD Optimizer Usage

### DISCUSSION THREAD

**Discussion by Brayan532** [ultralytics/ultralytics#23595] [Q&A] [answered]
Hi Ultralytics team, 

First of all, thank you for releasing YOLO26 and all the significant advancements you’ve brought to computer vision workflows. 
I’ve read the official YOLO26 documentation and blog posts that highlight MuSGD as a hybrid optimizer combining SGD with Muon-style optimization to drive stable training. However, I couldn’t find concrete examples or recommended usage patterns (such as code snippets, parameter settings, or API usage) in the documentation that show how to enable and configure MuSGD in user workflows, especially for custom models. This information would be extremely helpful for the community. 

Thanks very much for your continued investment in the Ultralytics ecosystem and for making cutting-edge CV technology more accessible!

Sincerely,
Brayan.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Brayan532, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get unblocked quickly—an Ultralytics engineer will also take a look and assist soon. We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report (for example, if MuSGD isn’t behaving as expected or you’re seeing errors enabling it), please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it (including your command/script, a small config snippet, and the exact logs).

If this is a custom training ❓ Question (like recommended MuSGD usage patterns, configuration knobs, or best-practice settings for custom models), please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) ✅.

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply, marked as answer]
Hi @Brayan532 — you can enable MuSGD in standard Ultralytics training by setting `optimizer="MuSGD"` (or leaving `optimizer="auto"`, which will select MuSGD for longer runs based on total iterations); for the exact behavior (which params get Muon vs plain SGD, and the internal Muon/SGD blend used), see the [MuSGD/Muon API reference](https://docs.ultralytics.com/reference/optim/muon/) and the trainer’s [optimizer construction logic](https://docs.ultralytics.com/reference/engine/trainer/#method-ultralyticsenginetrainerbasetrainerbuild_optimizer).

```python
from ultralytics import YOLO

model = YOLO("yolo26n.pt")
model.train(data="coco8.yaml", epochs=100, imgsz=640, optimizer="MuSGD")
```

If you’re training a fully custom PyTorch loop (not `model.train()`), you can also import `MuSGD` directly from `ultralytics.optim.muon` and pass explicit param groups (with `use_muon=True` only on 2D+ tensors), as shown in the same [MuSGD/Muon API reference](https://docs.ultralytics.com/reference/optim/muon/); feel free to reply here with your model type (detect/seg/pose) and whether you’re fine-tuning or training from scratch if you want a recommended starting `lr0`/`momentum` for your setup.

**Reply by Brayan532** [reply]
Dear Glenn, 
I'm grateful for your swift and clear reply. 

Best wishes, 
B.

**Reply by glenn-jocher** [reply]
Thanks Brayan—glad that clarified it, and credit to the broader YOLO community and the Ultralytics team; if you want feedback on MuSGD settings for your specific dataset/model (fine-tune vs scratch), just reply here with your `model.train(...)` args and a short snippet of your train logs (and if anything errors, please include the full traceback plus `yolo checks` output).
